anomalib/utils/callbacks/visualizer_callback.py

```diff
@@ -96,6 +95,60 @@ def _add_images(
         if "local" in module.hparams.project.log_images_to:
             visualizer.save(Path(module.hparams.project.path) / "images" / filename.parent.name / filename.name)
 
+    def generate_visualizer(self, outputs):
+        """Yields a visualizer object for each of the images in the output."""
+        for i in range(outputs["image"].size(0)):
+            visualizer = Visualizer()
+
+            image = Denormalize()(outputs["image"][i].cpu())
+            anomaly_map = outputs["anomaly_maps"][i].cpu().numpy()
+            heat_map = superimpose_anomaly_map(anomaly_map, image, normalize=not self.inputs_are_normalized)
+            pred_score = outputs["pred_scores"][i].cpu().numpy()
+            pred_label = outputs["pred_labels"][i].cpu().numpy()
+
+            if self.task == "segmentation":
+                pred_mask = outputs["pred_masks"][i].squeeze().int().cpu().numpy() * 255
+                vis_img = mark_boundaries(image, pred_mask, color=(1, 0, 0), mode="thick")
+                visualizer.add_image(image=image, title="Image")
+                if "mask" in outputs:
+                    true_mask = outputs["mask"][i].cpu().numpy() * 255
+                    visualizer.add_image(image=true_mask, color_map="gray", title="Ground Truth")
+                visualizer.add_image(image=heat_map, title="Predicted Heat Map")
+                visualizer.add_image(image=pred_mask, color_map="gray", title="Predicted Mask")
+                visualizer.add_image(image=vis_img, title="Segmentation Result")
+            elif self.task == "classification":
+                visualizer.add_image(image, title="Image")
+                if pred_label:
+                    image_classified = add_anomalous_label(heat_map, pred_score)
+                else:
+                    image_classified = add_normal_label(heat_map, 1 - pred_score)
+                visualizer.add_image(image=image_classified, title="Prediction")
+
+            yield visualizer
+
+    def on_predict_batch_end(
+        self,
+        _trainer: pl.Trainer,
+        _pl_module: AnomalyModule,
+        outputs: Optional[STEP_OUTPUT],
+        _batch: Any,
+        _batch_idx: int,
+        _dataloader_idx: int,
+    ) -> None:
+        """Show images at the end of every batch.
+
+        Args:
+            _trainer (Trainer): Pytorch lightning trainer object (unused).
+            _pl_module (LightningModule): Lightning modules derived from BaseAnomalyLightning object as
+            currently only they support logging images.
+            outputs (Dict[str, Any]): Outputs of the current test step.
+            _batch (Any): Input batch of the current test step (unused).
+            _batch_idx (int): Index of the current test batch (unused).
+            _dataloader_idx (int): Index of the dataloader that yielded the current batch (unused).
+        """
+        for visualizer in self.generate_visualizer(outputs):
+            visualizer.show()
```

Comment:
Since this is work in progress I am assuming the final design will not always show the results by default.